kernel: arch: introduce k_float_enable()

katsuster opened the pull request:
This patch introduce new API to enable FPU of thread. This is pair of
existed k_float_disable() API. And also add empty arch_float_enable()
into each architectures that have arch_float_disable(). The arc and
riscv already implemented arch_float_enable() so I do not touch
these implementations.

Motivation: Current Zephyr implementation does not allow to use FPU
on main and other system threads like as work queue. Users need to
create an other thread with K_FP_REGS for floating point programs.
Users can use FPU more easily if they can enable FPU on running
threads.

katsuster committed:
kernel: arch: introduce k_float_enable()

This patch introduce new API to enable FPU of thread. This is pair of
existed k_float_disable() API. And also add empty arch_float_enable()
into each architectures that have arch_float_disable(). The arc and
riscv already implemented arch_float_enable() so I do not touch
these implementations.

Motivation: Current Zephyr implementation does not allow to use FPU
on main and other system threads like as work queue. Users need to
create an other thread with K_FP_REGS for floating point programs.
Users can use FPU more easily if they can enable FPU on running
threads.

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

katsuster commented:
checkpatch got angry if I used ENOSYS in arch_float_enalbe() stub functions...

ioannisg commented on arch/arc/core/thread.c:
```diff
@@ -231,7 +231,6 @@ int arch_float_disable(struct k_thread *thread)
 	return 0;
 }
 
-
```
you could discard this white space diff :)

katsuster replied:
Oops... fixed.

ioannisg commented on arch/arm/core/aarch32/thread.c:
```diff
@@ -353,6 +353,11 @@ int arch_float_disable(struct k_thread *thread)
+
+int arch_float_enable(struct k_thread *thread)
+{
+	return -ENOSYS;
```
That's fine for me. Please, add an inline comment that "this is not supported in Cortex-M" or something similar

ioannisg replied:
FYI: Cortex-R does not support FPU in general, at the moment.

katsuster replied:
OK, added such information.

ioannisg commented on arch/riscv/core/thread.c:
```diff
@@ -151,7 +151,6 @@ int arch_float_disable(struct k_thread *thread)
 	return 0;
 }
 
-
```
same here, pls, revert this white space diff

katsuster replied:
Thanks. fixed.

ioannisg commented on kernel/include/kernel_arch_interface.h:
```diff
@@ -180,6 +180,17 @@ void arch_switch_to_main_thread(struct k_thread *main_thread, char *stack_ptr,
+ *
+ * The function is used to enable the preservation of floating
+ * point context information for a particular thread.
+ *
```
Add some text, also, here, that this API is optional , and depends on the architecture.

katsuster replied:
I added such text.

ioannisg reviewed (commented):
Seems to be in the right direction.

katsuster commented:
Conflicting k_float_enable() on x86 platform... I will check.

katsuster commented:
Changes in v2.
  - x86/ia32: change k_float_enable() -> z_float_enable() and call from arch_float_enable() same as arch_float_disable()
  - kernel: add "options" argument into k_float_enable()

dcpleung commented on arch/x86/core/ia32/float.c:
```diff
@@ -179,7 +179,7 @@ static inline void FpCtxInit(struct k_thread *thread)
  * thread (since context switching can't occur), but it is harmless.
  */
-void k_float_enable(struct k_thread *thread, unsigned int options)
+void z_float_enable(struct k_thread *thread, unsigned int options)
```
Renaming it to `arch_float_enable()` will get rid of the `z_float_enable()` redirection.

katsuster replied:
Thanks for your review. Yes, right. This is to keep consistency with existed k_float_disable() implementation.
If change such k_float_*() redirection of x86, it's better to change in an other commit (or PR ?).

dcpleung replied:
It's fine to update it in another PR.

ioannisg commented:
@katsuster pls, take a look at the compliance check failures

katsuster commented:
@ioannisg 
Thanks for comments. It seems scripts/checkpatch.pl always warned if we use ENOSYS in the C code.
I think we can ignore this compliance check error at this time.

pabigot commented:
Regarding `-ENOSYS` #29806 and particularly #23727 may be relevant.

I have no position on this PR nor on whether the checkpatch diagnostics on `-ENOSYS` should be removed.  Zephyr has been moving (if not entirely consistently) away from blindly following Linux coding standards, so any solution should be considered.

pabigot commented:
Based on the information provided  in #23727 I believe the Zephyr standard solution would be to use `-ENOTSUP` instead of `-ENOSYS` for the return value, avoiding the checkpatch error which is useful as it identifies cases where the return value is not consistent with Zephyr standard.  (The diagnostic message could be improved, though.)

carlescufi commented:
@katsuster we concluded that you should replace `ENOSYS` with `ENOTSUP`.
https://github.com/zephyrproject-rtos/zephyr/issues/23727#issuecomment-805051697

katsuster commented:
@pabigot Thanks for describing about ENOSYS and ENOTSUP.
@carlescufi Thanks a lot. I understand. I will fix my patch.

pabigot commented:
#33637 updates the checkpatch behavior.

katsuster committed:
kernel: arch: use ENOTSUP instead of ENOSYS in k_float_disable()

This patch replaces ENOSYS into ENOTSUP to keep consistency with
the return value specification of k_float_enable().

Signed-off-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

nashif commented:
@ruuddw  @evgeniy-paltsev @abrodkin  please review for arc

ruuddw commented:
> @ruuddw @evgeniy-paltsev @abrodkin please review for arc

The PR said "The arc and riscv already implemented arch_float_enable() so I do not touch these implementations." so I didn't pay much attention, thanks for explicitly asking. I see an "option" parameter has been added (used by x86), that's fine.
